ddsenabler_participants/src/cpp/EnablerParticipant.cpp

```diff
@@ -53,22 +90,84 @@ std::shared_ptr<IReader> EnablerParticipant::create_reader(
         return std::make_shared<BlankReader>();
     }
 
-    std::shared_ptr<InternalReader> reader;
+    std::shared_ptr<IReader> reader;
     {
         std::lock_guard<std::mutex> lck(mtx_);
-        reader = std::make_shared<InternalReader>(id());
         auto dds_topic = dynamic_cast<const DdsTopic&>(topic);
-        readers_[dds_topic] = reader;
-        // Only notify the discovery of topics that do not originate from a topic query callback
-        if (dds_topic.topic_discoverer() != this->id())
+        RpcInfo rpc_info = RpcUtils::get_rpc_info(dds_topic.m_topic_name);
+        if (RPC_TYPE::RPC_NONE != rpc_info.rpc_type)
+        {
+            if (SERVICE_TYPE::SERVICE_NONE != rpc_info.service_type)
+            {
+                reader = std::make_shared<InternalRpcReader>(id(), dds_topic);
+            }
+            else
+            {
+                reader = std::make_shared<InternalReader>(id());
+            }
+
+            // Only notify the discovery of topics that do not originate from a topic query callback
+            if (dds_topic.topic_discoverer() != this->id())
+            {
+                if (ACTION_TYPE::ACTION_NONE == rpc_info.action_type)
+                {
+                    if (service_discovered_nts_(rpc_info, dds_topic))
+                    {
+                        RpcTopic service = services_.find(rpc_info.service_name)->second->get_service();
+                        std::static_pointer_cast<Handler>(schema_handler_)->add_service(service);
+                    }
+                }
+                else
```

Comment:
Try to have explicit cases, will make easier to avoid errors if any of these enumerations are extended in the future.